openapi3filter/middleware.go

```diff
@@ -0,0 +1,272 @@
+// prescribing a particular one.
+func OnLog(f LogFunc) ValidatorOption {
+	return func(v *Validator) {
+		v.logFunc = f
+	}
+}
+
+// Strict, if set, causes an internal server error to be sent if the wrapped
+// handler response fails response validation. If not set, the response is sent
+// and the error is only logged.
+func Strict(strict bool) ValidatorOption {
+	return func(v *Validator) {
+		v.strict = strict
+	}
+}
+
+// Middleware returns an http.Handler which wraps the given handler with
+// request and response validation.
+func (v *Validator) Middleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		route, pathParams, err := v.router.FindRoute(r)
+		if err != nil {
+			v.logFunc("validation error: failed to find route for "+r.URL.String(), err)
+			v.errFunc(w, http.StatusNotFound, ErrCodeCannotFindRoute, err)
+			return
+		}
+		requestValidationInput := &RequestValidationInput{
+			Request:    r,
+			PathParams: pathParams,
+			Route:      route,
+		}
+		err = ValidateRequest(r.Context(), requestValidationInput)
+		if err != nil {
+			v.logFunc("invalid request", err)
+			v.errFunc(w, http.StatusBadRequest, ErrCodeRequestInvalid, err)
+			return
+		}
+
+		var wr responseWrapper
+		if v.strict {
+			wr = &strictResponseWrapper{w: w}
+		} else {
+			wr = &warnResponseWrapper{w: w}
+		}
+
+		h.ServeHTTP(wr, r)
+
+		err = ValidateResponse(r.Context(), &ResponseValidationInput{
+			RequestValidationInput: requestValidationInput,
+			Status:                 wr.statusCode(),
+			Header:                 wr.Header(),
+			Body:                   nopCloser{bytes.NewBuffer(wr.bodyContents())},
+		})
+		if err != nil {
+			v.logFunc("invalid response", err)
+			if v.strict {
+				v.errFunc(w, http.StatusInternalServerError, ErrCodeResponseInvalid, err)
+			}
+			return
+		}
+
+		err = wr.flushBodyContents()
+		if err != nil {
+			v.logFunc("failed to write response", err)
+		}
+	})
+}
+
+type nopCloser struct {
+	io.Reader
+}
+
+// Close implements io.Closer.
+func (nopCloser) Close() error {
+	return nil
+}
+
+type responseWrapper interface {
+	http.ResponseWriter
+
+	// flushBodyContents writes the buffered response to the client, if it has
+	// not yet been written.
+	flushBodyContents() error
+
+	// statusCode returns the response status code, 0 if not set yet.
+	statusCode() int
+
+	// bodyContents returns the buffered
+	bodyContents() []byte
+}
+
+type warnResponseWrapper struct {
+	w      http.ResponseWriter
+	fl     http.Flusher
+	status int
+	body   bytes.Buffer
+}
+
+// Write implements http.ResponseWriter.
+func (l *warnResponseWrapper) Write(b []byte) (int, error) {
+	if l.status == 0 {
+		l.w.WriteHeader(l.status)
+	}
+	n, err := l.w.Write(b)
+	if err == nil {
+		l.body.Write(b)
+	}
+	return n, err
+}
+
+// WriteHeader implements http.ResponseWriter.
+func (l *warnResponseWrapper) WriteHeader(status int) {
+	if l.status == 0 {
+		l.status = status
+	}
```

Comment:
Why compare to the empty value when `status` is here?
Reading this makes me think that zero is some special case and `l.status` may have been written to beforehand and so its value should only be overwritten in that case... but it's not clear to me what can mutate `l.status` appart from this place.

This surely alludes to my comment on 
```go
func (l *warnResponseWrapper) Write(b []byte) (int, error) {
	if l.status == 0 {
		l.w.WriteHeader(l.status)
	}
```

In any case, I think you should keep assigning here but drop the conditional; remove the call to `WriteHeader` in `Write`

Reply:
I think separating the status we record, from the state of whether we've "written the header" (or captured it) should hopefully clear this up, updated.